[GStreamer] Rewrite inner decoder leveraging the GStreamerElementHarness

philn opened the pull request:
#### fa3503d135d0820bcdbb68e51516ebfd75fd090d
<pre>
[GStreamer] Rewrite inner decoder leveraging the GStreamerElementHarness
<a href="https://bugs.webkit.org/show_bug.cgi?id=250934">https://bugs.webkit.org/show_bug.cgi?id=250934</a>

Reviewed by Xabier Rodriguez-Calvar.

The previous decoder implementation relied on a GStreamer pipeline involving giostreamsrc,
decodebin3 and appsinks. Due to the highly asynchronous behavior of that pipeline the decoder
heavily relied on Locks and Conditions making the code hard to follow and debug.

The new implementation relies on two element harnesses, the first one is in charge of parsing (with
parsebin), the first parsed video output stream is then plugged to a decode harness, directly hooked
to the platform video decoder, so no queue is involved. The decoding no longer involves additional
threads, so we can simplify the code and entirely get rid of InnerDecoder.

Canonical link: <a href="https://commits.webkit.org/259430@main">https://commits.webkit.org/259430@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/50b69133b0213ad646470401b46c50c31c0237b5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104718 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13795 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37621 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113993 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174195 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108634 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/14923 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4726 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97049 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112919 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110478 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11516 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94546 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39072 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/108187 "Passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93376 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26160 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80739 "Passed tests") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/94648 "Found 1 new JSC stress test failure: wasm.yaml/wasm/lowExecutableMemory/exports-oom.js.default-wasm (failure)") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7149 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27516 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/92610 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/4898 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7254 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4095 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30169 "Passed tests") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/20/builds/103539 "Passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13303 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47075 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/101297 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6507 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9042 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25239 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/platform/graphics/gstreamer/ImageDecoderGStreamer.cpp:
```diff
@@ -228,81 +232,56 @@ const ImageDecoderGStreamerSample* ImageDecoderGStreamer::sampleAtIndex(size_t i
-    gst_element_sync_state_with_parent(videoconvert);
-    gst_element_sync_state_with_parent(sink);
+        GRefPtr<GstElement> element = gst_element_factory_create(lookupResult.factory.get(), nullptr);
+        // TODO: This was copied from our WebCodec video decoder, it should be refactored in a dedicated function.
```
Per style, we only have FIXMEs

calvaris commented on Source/WebCore/platform/graphics/gstreamer/ImageDecoderGStreamer.cpp:
```diff
@@ -228,81 +232,56 @@ const ImageDecoderGStreamerSample* ImageDecoderGStreamer::sampleAtIndex(size_t i
+        GUniquePtr<char> elementName(gst_element_get_name(element.get()));
+        auto elementHasProperty = [&](const char* name) -> bool {
+            return g_object_class_find_property(G_OBJECT_GET_CLASS(element.get()), name);
+        };
```
Maybe we can refactor this in a function in GStreamerCommon

calvaris commented on Source/WebCore/platform/graphics/gstreamer/ImageDecoderGStreamer.h:
```diff
@@ -78,47 +79,32 @@ class ImageDecoderGStreamer final : public ImageDecoder {
-        Condition m_messageCondition;
-        Lock m_messageLock;
-        bool m_messageDispatched WTF_GUARDED_BY_LOCK(m_messageLock) { false };
+        Span<const uint8_t> m_buffer;
```
I think you should turn this into a Vector. A Span is something not owned and from what I see, you store it in the constructor and use it in the `run` method where you turn it indeed into a Vector. Between the call to the constructor and `run`, it is not clear what it can happen to the ownership of that data, so better to be safe than sorry.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/ImageDecoderGStreamer.cpp:
```diff
@@ -318,131 +297,63 @@ void ImageDecoderGStreamer::notifySample(GRefPtr<GstSample>&& sample)
     m_eos = false;
-    auto thread = Thread::create("ImageDecoderGStreamer", [this, data = contiguousBuffer->data(), size = buffer.size()] {
-        m_innerDecoder = ImageDecoderGStreamer::InnerDecoder::create(*this, data, size);
+    auto thread = Thread::create("ImageDecoderGStreamer", [this, buffer = Span<const uint8_t> { data.data(), data.size() }] {
```
I guess you're assuming that this this lambda you pass to `Thread::create` is going to stay alive, probably with the `thread`, at least until it is run because otherwise you'd need to not use a `Span` and take ownership of that data into to the lambda.

calvaris reviewed (changes requested):
Other than the changes I also commented, I think it could be a good idea to change the interface of `pushBuffer` and `pushBufferFull` for them to take a `GRefPtr&&` that gives an explicit view of ownership. Maybe even all push* functions could use a similar interface.

The downside of this is that you have to WTFMove objects and adopt the ones with _[transfer full]_ before passing them down, or we'd be leaking.

About this I don't have a strong opinion so I just leave it here as "food for thought".

philn commented on Source/WebCore/platform/graphics/gstreamer/ImageDecoderGStreamer.cpp:
```diff
@@ -318,131 +297,63 @@ void ImageDecoderGStreamer::notifySample(GRefPtr<GstSample>&& sample)
+    Vector<uint8_t> data { m_buffer };
+    auto bufferSize = data.size();
+    auto bufferData = data.data();
+    auto buffer = gst_buffer_new_wrapped_full(GST_MEMORY_FLAG_READONLY, bufferData, bufferSize, 0, bufferSize, new Vector<uint8_t>(WTFMove(data)), [](gpointer data) {
```
Actually, reading this again, I suspect we don't need a vector here at all. And no need to make a copy either, because the buffer is passed right away to the harness. I'll revisit this. Thanks for the review!

calvaris replied:
If you keep the constructor + `run` scheme, as I mention above, you should turn the span into a vector in the constructor already, otherwise the data ownership is not clear. And also the Thread + lambda thing.

philn replied:
The data is safe because we wait the decoding completion anyway.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
```diff
@@ -1041,6 +1041,19 @@ void fillVideoInfoColorimetryFromColorSpace(GstVideoInfo* info, const PlatformVi
 
+void configureVideoDecoderForHarnessing(const GRefPtr<GstElement>& element)
+{
+    auto elementHasProperty = [&](const char* name) -> bool {
```
It is not the first time I see this, please strongly consider moving this to GStreamerCommon and refactor all appearances. It does not have to be in this PR though.

philn committed:
[GStreamer] Rewrite inner decoder leveraging the GStreamerElementHarness
https://bugs.webkit.org/show_bug.cgi?id=250934

Reviewed by Xabier Rodriguez-Calvar.

The previous decoder implementation relied on a GStreamer pipeline involving giostreamsrc,
decodebin3 and appsinks. Due to the highly asynchronous behavior of that pipeline the decoder
heavily relied on Locks and Conditions making the code hard to follow and debug.

The new implementation relies on two element harnesses, the first one is in charge of parsing (with
parsebin), the first parsed video output stream is then plugged to a decode harness, directly hooked
to the platform video decoder, so no queue is involved. The decoding no longer involves additional
threads, so we can simplify the code and entirely get rid of InnerDecoder.

Canonical link: https://commits.webkit.org/259430@main
